internal/services/mongocluster/mongo_cluster_resource.go

```diff
@@ -0,0 +1,572 @@
+	PublicNetworkAccessEnabled bool              `tfschema:"public_network_access_enabled"`
+	PreviewFeatures            []string          `tfschema:"preview_features"`
+	StorageSizeInGb            int64             `tfschema:"storage_size_in_gb"`
+	Tags                       map[string]string `tfschema:"tags"`
+	Version                    string            `tfschema:"version"`
+}
+
+func (r MongoClusterResource) ModelObject() interface{} {
+	return &MongoClusterResourceModel{}
+}
+
+func (r MongoClusterResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return mongoclusters.ValidateMongoClusterID
+}
+
+func (r MongoClusterResource) ResourceType() string {
+	return "azurerm_mongo_cluster"
+}
+
+func (r MongoClusterResource) Arguments() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			ForceNew: true,
+			Required: true,
+			Type:     schema.TypeString,
+			ValidateFunc: validation.StringMatch(
+				regexp.MustCompile(`^[a-z\d]([-a-z\d]{1,38}[a-z\d])$`),
+				"`name` must be between 3 and 40 characters. It can contain only lowercase letters, numbers, and hyphens (-). It must start and end with a lowercase letter or number.",
+			),
+		},
+
+		"resource_group_name": commonschema.ResourceGroupName(),
+
+		"location": commonschema.Location(),
+
+		"administrator_username": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			RequiredWith: []string{"administrator_username", "administrator_password"},
+		},
+
+		"create_mode": {
+			Type:     schema.TypeString,
+			Optional: true,
+			ForceNew: true,
+			Default:  string(mongoclusters.CreateModeDefault),
+			// Confirmed with service team the 'Default' and `GeoReplica` are the only accepted value currently, other values will be supported later.
+			ValidateFunc: validation.StringInSlice([]string{
+				string(mongoclusters.CreateModeDefault),
+				string(mongoclusters.CreateModeGeoReplica),
+			}, false),
+		},
+
+		"preview_features": {
+			Type:     schema.TypeList,
+			Optional: true,
+			ForceNew: true,
+			Elem: &schema.Schema{
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice(mongoclusters.PossibleValuesForPreviewFeature(), false),
+			},
+		},
+
+		"shard_count": {
+			Type:         schema.TypeInt,
+			Optional:     true,
+			ValidateFunc: validation.IntAtLeast(1),
+			ForceNew:     true,
+		},
+
+		"source_location": {
+			Type:             schema.TypeString,
+			Optional:         true,
+			ForceNew:         true,
+			StateFunc:        location.StateFunc,
+			DiffSuppressFunc: location.DiffSuppressFunc,
+			ValidateFunc:     validation.StringIsNotEmpty,
+			RequiredWith:     []string{"source_server_id"},
+		},
+
+		"source_server_id": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```

Comment:
This should have a stricter validation 
```suggestion
			ValidateFunc: mongoclusters.ValidateMongoClusterID,
```

Reply:
Fixed.